### DISCUSSION TITLE
Workflow: Create Sarcophagus

### DISCUSSION THREAD

**Discussion by adamgall** [sarcophagus-org/sarcophagus-v2-specs#13] [Ideas]
This post is to talk through and understand the process of creating a new sarcophagus.

"Creating a new sarcophagus" means that an end user ("embalmer") wishes to create a dead man's switch payload, which a recipient should gain access to in the future, when the embalmer stops extending the switch ("re-wrapping the sarcophagus").

**Comment by adamgall** [comment]
This comment thread can contain the official workflow, as we work through the nuances via other comments.

First pass at this workflow:

1. embalmer arrives at webapp
2. embalmer indicates to webapp that they'd like to create a new sarcophagus, uploads raw data and recipient's public key
3. webapp makes connections to the archaeologist p2p network
4. webapp communicates with all found archaeologists, and gathers information about them (fee limits, bond amounts, reputation, "next public key", what else?). some of this data comes from archaeologists directly (fee limits, keys), some is gathered from the blockchain (bond amounts, reputation).
6. webapp chooses a set of `m` archaeologists to participate in this new sarcophagus
7. webapp generates public / private key pair for this new sarcophagus
8. webapp uses Shamir's Secret Sharing Scheme to create `n` shards for the newly generated private key
9. webapp hashes each shard and holds onto those hashes, including them into the bundle going to Arweave
10. webapp encrypts each shard with a chosen archaeologist's public key
11. webapp encrypts raw data with recipient's public key
12. webapp calculates hash of this encrypted data
13. webapp calculates hash of the just-calculated hash (this is called the "double hash" in V1 terminology)
14. webapp encrypts the encrypted data from step 11 with the newly generated public key from step 7
15. webapp creates bundle of data to be saved to Arweave, consisting of fully encrypted payload (two layers of encryption), along with all of the encrypted private key shards (and likely some identifier to tie a shard to a specific archaeologist)
16. webapp chooses an "arweave archaeologist" from the p2p network, gets their rates, figures out how much this bundle will cost (based on size and archaeologist rates)
17. webapp transmits bundle to the Arweave archaeologist
18. webapp makes first transaction to blockchain, creating a unique identifier for this new sarcophagus, and putting up payment for the to-be-completed Arweave transaction (as calculated from previous step), as well as payment for first round of wrapping, identifiers all archaeologists, and the public keys for each archaeologist used to encrypt their shards.
    * Does this transaction need to specify an Arweave archaeologist that should complete the Arweave transaction, or can it be fulfilled by any Arweave archaeologist?
19. Arweave archaeologist sees blockchain transaction, confirms that the payment they'll receive is sufficient, makes the Arweave transaction
20. Arweave archaeologist sends Arweave transaction ID back to Embalmer
21. webapp confirms that the data at the Arweave transaction is correct
22. webapp collects signatures from archaeologists, which they should only sign if they're on board with all of the information in the first blockchain transaction
23. webapp completes the sarcophagus creation by making another blockchain transaction which includes the Arweave transaction ID, and all archaeologist signatures

**Comment by dacebt** [comment]
> 5. webapp chooses a set of m archaeologists to participate in this new sarcophagus

I was thinking that we expand this workflow to have to options.
- Advanced Settings (choose each of the `m` archaeologist)
- Allow the system to choose automatically based on input from emblamer. Something like a Max Fees or something. and the system can select these archaeologist for the emblamer and present them with a 'auto' selection. that they can then change if they want to.

**Comment by USCMig** [comment]
Just capturing this as a note for the future, since this wasn't really brought up earlier, but in line with the person who tor-ified their Archaeologist node, a way to potentially preserve the "identity" of either the embalmer/recipient and/or the archaeologists involved in the creation of sarcophagus an implementation of a zk proof could be done in a zk circuit and then executed in that contract.  I'm thinking that this could be implemented in either steps 4/5 or step 9 dependent on whose identity we want to keep shielded from the other party.  This could reduce the chance of a mass node operator having control of a majority of the `m` of `n` arch nodes since they could be shielded.

I found this (albeit old) but valid (of sorts) example... https://kndrck.co/posts/practical_guide_build_zk_dapps/

We could maybe talk about this being a part of the next version since we already have a game plan set for v2 development but wanted to throw this out there.

**Reply by adamgall** [reply]
Thank you @USCMig , I agree that including more privacy-preserving tooling into Sarcophagus would be an amazing addition to the system.

The inclusion of ZK implementations into this system is a big discussion that I'm not sure I have the necessary background or context to have right now. Might be a good discussion topic on the next dev chat next week.
